ballerina/in_memory_datasource.bal

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2024 WSO2 LLC. (http://www.wso2.com) All Rights Reserved.
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql_schema_registry.datasource;
+
+type SupergraphSubgraph record {|
+    readonly string supergraphVersion;
+    readonly string subgraphName;
+    string subgraphVersion;
+|};
+
+public isolated client class InMemoryDatasource {
+    *datasource:Datasource;
+
+    private final table<datasource:Supergraph> key(version) supergraphTable;
+    private final table<datasource:Subgraph> key(version, name) subgraphTable;
+    private final table<SupergraphSubgraph> key(supergraphVersion, subgraphName) supergraphSubgraphTable;
+
+    public function init() {
+        self.supergraphTable = table [];
+        self.subgraphTable = table [];
+        self.supergraphSubgraphTable = table [];
+    }
+
+    isolated resource function get supergraphs() returns datasource:Supergraph[]|datasource:Error {
+        lock {
+            return self.supergraphTable.toArray().clone();
```

Comment:
That would require the output type of the function signature (`get supergraphs() returns datasource:Supergraph[]|datasource:Error`) to be changed right? Since In-Memory datasource uses the Datasource api, changing the signature is not possible right? Please correct me if I'm wrong 🥹